### ISSUE TITLE
State is not properly cleared in case of broken hooks

### ISSUE THREAD

**Issue by theKashey** [facebook/react#14881]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

With one hook rendered wrongly, with a fewer hooks for example, rest of the application is also broken.
All due _skipped_ render cleanup phase.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**

```js
 <AppWithErrorBoundary>
  <HooksInsideErrorBoundary />
  <AnyFunctionalComponentWithoutHooks />
 </AppWithErrorBoundary>
```
Then _break_ hooks.

- first error boundary called, and an error state is properly set.
- `AnyFunctionalComponentWithoutHooks` is rendered without `currentHook` being reset from the first render.
- `didRenderTooFewHooks` set to true
- error propagated to `AppWithErrorBoundary`

In other words - bad hook is poisoning everything behind it.

**What is the expected behavior?**

Expected behavior - `error` should be catched by HooksErrorBoundary.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

16.8.2-

**Comment by gaearon** [comment]
A reproducing case would help.

**Comment by theKashey** [comment]
I've tried to reproduce the problem, and look like it's impossible to break React in a normal case. https://codesandbox.io/s/91wpr2q6kr - always working, and I've done my best to crush it.

Look like the problem is bound to my case, where I am `forceUpdate`ing all components, and that causes hooks to be rendered twice, throw error twice, and thus bypass ErrorBoundary via `replayUnitOfWork`, which is leading to poisoning.
There is a mistake in my code, I should not call forceUpdate outside the React lifecycle.

If you still looking for a reproducible demo - https://github.com/gaearon/react-hot-loader/tree/master/examples/styled-components - just change the Counter and it would be broken.

**Comment by gaearon** [comment]
Calling forceUpdate seems legit. Can you still create an isolated repro case?

**Comment by theKashey** [comment]
I've found the root cause - it's cache busting. Literally - 
```
  fiber.memoizedProps = Object.assign(
    { cacheBusterProp: true },
    fiber.memoizedProps,
  )
```
That was added to _unmemoize_ memoize. And that's killing some things. Not killing - Error got propagated one ErrorBoundery upper.

**Comment by theKashey** [comment]
Here is a small demo to reproduce - https://codesandbox.io/s/m77545vjrj
- to break - comment out second `useState`. Then the whole application would fail.
- to fix - comment out `useEffect`. Then error would be bound to Hooked element
- or remove `AnotherComponent` which is killing the app due the poisoning I am talking about - it error would be bound to the Hooked element.

Look like the root cause is useEffect, but only in special conditions.

**Comment by gaearon** [comment]
If the error is only reproducible when patching internal fields then it seems like it's a low pri issue.

**Comment by theKashey** [comment]
It's more about - clean stack after exception. You are already doing it, but not for all code branches.
